v0.2.3: Spectrum readers, spectroelectrochemistry refactor, CyclicVoltammogram.plot_cycles

ScottSoren opened the pull request:
This PR includes a few small updates for an 0.2.3 release. They are described in NEXT_CHANGES.rst. Briefly:

- A reader for .xrdml files saved by e.g. PanAnalytical X-ray diffractometers ([example here](https://www.dropbox.com/s/yufmsj7qfyeh5pc/GI-XRD%20Path%202_1%20omega%200p5%20step%2010s.xrdml?dl=0)), whcih returns a plain Spectrum object
  This makes use of pythons `xml` library. I want to improve it after finishing [this XML tutorial](https://realpython.com/python-xml-parser/). For now I just got it to work as quick as possible. 
- A reader for .avg files exported from the Batch ThermoFisher's Avantage Software ([example here](https://www.dropbox.com/s/4tj3jhsuqk0zsxx/XPS%20Survey.avg?dl=0)). This is copied with the absolute minimum necessary modification from old code I had lying around.
- A new method called `CyclicVoltammogram.plot_cycles()` which makes beautiful colored stuff like [this figure](https://www.dropbox.com/s/310xufs2iwgjq7x/demo_01.png?dl=0). Nice selling point of ixdat to use in demos :) ... however it should probably be moved to a `Plotter`
- A minor debug of the PV Mass Spec ("pfeiffer") reader.

@Ejler , if you have time to look at src/ixdat/readers/xrdml.py and especially .../avantage.py, that would be great. I know you have more advanced code for reading the avantage files, which would be great to replace this code with in a future PR :) 

@KennethNielsen , if you could just quickly check that the code is readable, that would be great. It's probably not worth going in detail on the structure readers, which I know can be improved. But what is important, I think, is the way the `Spectrum` object is initialized.

It may be bad style, but I think these small additions are important for demo'ing ixdat, and so I had people install [a pre-release](https://pypi.org/project/ixdat/0.2.3.dev1/) with this unreviewed code for the DTU demo yesterday. That's why I'm PR'ing now even though I know the code could use improvements... and docs and tests are neglected to fall yet further behind.

KennethNielsen reviewed (changes requested):
Ok, so I finished. A couple general things. I couldn't quite figure out the expected level of the read, since I couldn't figure out whether someone else would read it/has read it thoroughly. I read it like I normally would, but didn't check the details of whether all the semantics make sense with regards to the objects used i.e. read through of the semantics mostly dealt with whether the code locally made sense.

That being said. A few general points:

1. I remember that you before used keyword-only arguments to great effect. I think that is a magnificent idea. Both because we sometimes change signatures and because it improves readability at call site. So maybe consider sprinkling some of that on. Especially e.g. I think more or less all plot functions should make all KW args KWO args.
2. As far as the DB diagram. Besides from some concerns about what is possible (which we will talk about later) I would suggest to make a calculated spelling error and call the table "spectrums", so that we can keep the id_columns naming conventions simple, as per my comments in the other PR. I know that misspelling on purpose will annoy some people, but in this case, it really is practical.
3. I think maybe you ran out of steam, but the code seems to be a little light on the doc strings on the user facing methods towards the end.

KennethNielsen commented on src/ixdat/readers/avantage.py:
```diff
@@ -0,0 +1,124 @@
+    def read(self, path_to_file, cls=None, **kwargs):
+        """Load data stored as text by Advantage's default exporting mode
+
+        Copied from pyThetaProbe, written by Anna Winiwarter and Soren Scott in 2019
```
But alas, I now see that it is not complete

ScottSoren replied:
Nice! Then lots of people are qualified to improve or replace the one here now. @Ejler too, for example ;)

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -11,3 +11,138 @@ links to relevant Issues, Discussions, and PR's on github with the following for
+  XAS data (the "FFIO" column) is returned::
+
+    from ixdat import Spectrum
+    all_the_stuff = Spectrum.read("my_file.dat", reader="qexafs")
```
Now `multi_spectrum`.

ScottSoren commented on development_scripts/reader_testers/test_qexafs_reader.py:
```diff
@@ -0,0 +1,34 @@
+from ixdat import Spectrum, Measurement
+
+data_dir = Path(
+    r"C:\Users\scott\Dropbox\WORKSPACES\Caiwu\22E23_beamline_data\constant potential"
```
Ah, yes, this is unpublished data. I'll talk with @Caiwu-L about whether there's an example data set we can use.

ScottSoren commented on src/ixdat/measurements.py:
```diff
@@ -245,26 +252,25 @@ def read_set(cls, path_to_file_start, reader, suffix=None, file_list=None, **kwa
+            part (Path or str): A path where the folder is the folder containing data
+                and the name is a part of the name of each of the files to be read and
+                combined.
+            suffix (str): If a suffix is given, only files with the specified ending are
```
fixed.

ScottSoren commented on src/ixdat/plotters/plotting_tools.py:
```diff
@@ -15,3 +18,24 @@ def color_axis(ax, color, lr="right", xy="y"):
+    return rgba
+
+
+def add_colorbar(ax, cmap_name, vmin, vmax, label="intensity"):
```
added one.

ScottSoren commented on src/ixdat/plotters/sec_plotter.py:
```diff
@@ -2,63 +2,54 @@
-    FIXME: This should make use of the code in spectrum_plotter.SpectrumSeriesPlotter
-    """
+class SECPlotter(SpectroMeasurementPlotter):
+    """An spectroelectrochemsitry (SEC) matplotlib plotter."""
```
Wow, very consistent typo :P . It might be easier for the dictionary to just correct it to my spelling ;) 
I'll make an effort to start pronouncing it that way too.

ScottSoren commented on src/ixdat/plotters/sec_plotter.py:
```diff
@@ -100,63 +89,153 @@ def plot_measurement(
+            **kwargs,
+        )
+
+        axes[0] = super().heat_plot_vs(
```
Not needed, and I don't think it helps readability, given that the ec_plotter didn't assign back to axes[1] in the previous statement. I've removed this.

ScottSoren commented on src/ixdat/plotters/sec_plotter.py:
```diff
@@ -168,50 +247,90 @@ def plot_vs_potential(
+            ax (matplotlib Axis): The axes to plot on. A new one is made by default.
+            V_ref (float): potential to use as reference for calculating optical density
+            t_ref (float): time to use as a reference for calculating optical density
+            cmap_name (str): The name of the colormap to use. Defaults to "inferno",
```
Matplotlib's "inferno" seems to be in a "good" category by the criterium you mention. According to their website, it is a "perceptually uniform sequential color map": https://matplotlib.org/3.5.0/tutorials/colors/colormaps.html

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -61,107 +62,69 @@ def heat_plot(
 
         Args:
-            spectrum_series (SpectrumSeries): The spectrum series to be plotted, if
+            measurement (SpectrumSeries): The spectrum series to be plotted, if
```
Oh yeah, and the paragraph above the Args list is way out of sync. I've fixed it now.

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
-                FIXME: ...series to plot against if vs isn't in field.series_axes
             field (Field): The field to be plotted, if different from
                 spectrum_series.field
             ax (matplotlib Axis): The axes to plot on. A new one is made by default.
```
Fixed. Also, kwargs-only now :)

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+                different from self.spectrum_series.
+            field (Field): The field to be plotted, if different from
+                spectrum_series.field
+            xspan (iterable): The span of the spectral data to plot
```
Added. Also moved `vs` up the argument list since it's more central to this plotting method.

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+        self.spectrum_series_plotter = SpectrumSeriesPlotter()
+
+    def heat_plot_vs(
+        self,
```
yes. All the Spectrum and SpectroMeasurement plotting functions are now kwarg-only :)

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+        else:
+            v = measurement.grab_for_t(vs, t=tseries.t)
+
+        return self.spectrum_series_plotter.heat_plot(
```
fixed.

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+        """Plot a SpectrumSeries as spectra colored by the value at which they are taken
+
+        Args:
+            spectrum_series (SpectrumSeries): The spectrum series to be plotted, if
```
fixed.

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+                different from self.spectrum_series.
+            field (Field): The field to be plotted, if different from
+                spectrum_series.field
+            ax (matplotlib Axis): The axes to plot on. A new one is made by default.
```
fixed.

ScottSoren commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+            field (Field): The field to be plotted, if different from
+                spectrum_series.field
+            ax (matplotlib Axis): The axes to plot on. A new one is made by default.
+            cmap_name (str): The name of the colormap to use. Defaults to "inferno",
```
yes.

ScottSoren commented on src/ixdat/readers/avantage.py:
```diff
@@ -0,0 +1,124 @@
+        """Load data stored as text by Advantage's default exporting mode
+
+        Copied from pyThetaProbe, written by Anna Winiwarter and Soren Scott in 2019
+        This is also that used by the "TXT" batch exporter.
```
More explicit now:
The .avg format is also the format produced by the "TXT" batch exporter.

ScottSoren commented on src/ixdat/readers/avantage.py:
```diff
@@ -0,0 +1,124 @@
+        )
+        header_lines = []
+
+        with open(path_to_file, "r") as f:
```
I've added a TODO with a link to this comment, the PyExpLabSys reader, and Jacob's reader in LowOverpotentialRegime

ScottSoren commented on src/ixdat/spectra.py:
```diff
@@ -253,6 +289,109 @@ def __add__(self, other):
+        self._xseries = None
+        self._spectra = None
+
+    @property
```
Oh, you once again caught my sloppy not-propery-tested coding. I've fixed up the method and added some comments. But the mistakes I can see had to do with the assert being in the wrong place and a forgotten `_`, and don't actually address your specific question. To that:

`_xseries` gets set if the property `fields` is called. All the fields have to thave the same or equivalent xseries. So the property `xseries` did indeed work like the typical cache. I've rewritten it though to be explicit.

KennethNielsen replied:
Ah ok. I think my main trigger was the part where the cache was being written on in two methods. That is now gone, but what I'm still (subjectively) slightly wondering is if it wouldn't be cleaner with:
```python
    def fields(self):
        """Make sure Fields are loaded and have the same xseries"""
        xseries = self.xseries
        for i, f in enumerate(self._fields):
            if isinstance(f, PlaceHolderObject):
                # load or "unpack" any fields for which only the id's were loaded:
                self._fields[i] = f.get_object()
                # If all the xseries are the same, every field shoud have an xseries identical to the first one
                # retrieved via self.xseries
                assert self._fields[i].axes_series[0] == xseries
        # Now we've loaded any place-holder fields and checked their xseries are equal.
        return self._fields
```

ScottSoren replied:
That does look cleaner, but the problem is that the `self.xseries` property calls `self.fields`, so it would be an infinite recursion.

ScottSoren commented on src/ixdat/spectra.py:
```diff
@@ -253,6 +289,109 @@ def __add__(self, other):
+    extra_linkers = {"multispectrum_fields": {"data_series", "f_ids"}}
+    child_attrs = ["fields"]
+
+    def __init__(
```
yes. added one now.

ScottSoren commented on src/ixdat/spectra.py:
```diff
@@ -358,3 +525,83 @@ def __getitem__(self, key):
+        return self._spectrum_series
+
+    @property
+    def spec_id(self):
```
Renamed spectrum_id

ScottSoren commented on src/ixdat/spectra.py:
```diff
@@ -253,6 +289,109 @@ def __add__(self, other):
+        return self._xseries or self._fields[0].xseries
+
+    @property
+    def spectrum_list(self):
```
I've added some brief ones now throughout the module.

ScottSoren commented on src/ixdat/readers/qexafs.py:
```diff
@@ -0,0 +1,91 @@
+                    )
+
+        df = pd.read_csv(path_to_file, sep="\t", header=i - 1)
+        df.columns = df.columns.str.strip("#")
```
yeah, that's exactly what it does. It's wierd magic pandas syntax. I googled my way to it. But the file is dumbly formatted such that  the first column gets a name starting with a "# " without this stripping.
I'll add a comment.

KennethNielsen replied:
But I think what I was more wondering about was the difference in calling semantics. One line seemed to assign back and the next didn't, but I see that the second line now is different from what I remembered.

ScottSoren commented on src/ixdat/readers/avantage.py:
```diff
@@ -0,0 +1,124 @@
+                        )
+                    in_data = False
+                else:
+                    y_vec = np.append(y_vec, y_i)
```
As mentioned, to be improved in a later PR.

ScottSoren commented:
Hi @KennethNielsen , thanks for the review! As usual, you caught a lot of mistakes and omissions and the code improved as a result. I've responded to each of the comments and only left a few unresolved where I couldn't immediately implement the suggestions.
I completely agree with your top-level comments as well. I've made all the spectrum and spectromesurement plotting functions key-word only, and I've changed the table name from "spectra" to "spectrums". I added a note on grammer in the spectra.py module docstring justifying the use of two distinct plurals for spectrum. I've updated the dbdiagram: 
https://dbdiagram.io/d/629f3b5b54ce2635277667f5
Let me know if it looks good to you.

KennethNielsen commented on src/ixdat/plotters/spectrum_plotter.py:
```diff
@@ -170,53 +133,162 @@ def plot_waterfall(
+
+        Args:
+            measurement (SpectrumSeries): The spectrum series to be plotted, if
+                different from self.spectrum_series.
```
I was supposed to say "quite a bit better", but I mean it, it reads nicer and more fluent :D

KennethNielsen reviewed (approved):
Looks great. There is one comment that I unresolved, but with a subjective comment, so you decide whether you think it is better. No-matter what, I consider it ready for merge on my account.

Caiwu-L reviewed (commented):
The qexafs reader looks great to me, haven't find any problems. Using QexafsFFI0 as y when calling technique 'XAS' is right for sample measuring under fluorescence mode. Fluorescence mode have been commonly used in operando EC-XAS experiment. There is also a transition mode which is usually used to measure ex-situ standard sample, in such case, the 'XAS' should be lnIt/I0.
